orchagent/fdborch.cpp

```diff
@@ -118,85 +128,233 @@ void FdbOrch::update(sai_fdb_event_t type, const sai_fdb_entry_t* entry, sai_obj
     FdbUpdate update;
     update.entry.mac = entry->mac_address;
     update.entry.bv_id = entry->bv_id;
+    update.type = "dynamic";
+    Port vlan;
 
     switch (type)
     {
     case SAI_FDB_EVENT_LEARNED:
+    {
+        SWSS_LOG_INFO("Received LEARN event for bvid=%lx mac=%s port=%lx", entry->bv_id, update.entry.mac.to_string().c_str(), bridge_port_id);
+
+        if (!m_portsOrch->getPort(entry->bv_id, vlan))
+        {
+            SWSS_LOG_NOTICE("FdbOrch LEARN notification: Failed to locate vlan port from bv_id 0x%lx", entry->bv_id);
+            return;
+        }
+
         if (!m_portsOrch->getPortByBridgePortId(bridge_port_id, update.port))
         {
-            SWSS_LOG_ERROR("Failed to get port by bridge port ID 0x%" PRIx64, bridge_port_id);
+            SWSS_LOG_NOTICE("FdbOrch LEARN notification: Failed to get port by bridge port ID 0x%lx", bridge_port_id);
             return;
         }
 
         // we already have such entries
-        if (m_entries.find(update.entry) != m_entries.end())
+        auto existing_entry = m_entries.find(update.entry);
+        if (existing_entry != m_entries.end())
         {
-             SWSS_LOG_INFO("FdbOrch notification: mac %s is already in bv_id 0x%" PRIx64,
-                    update.entry.mac.to_string().c_str(), entry->bv_id);
+             SWSS_LOG_INFO("FdbOrch LEARN notification: mac %s is already in bv_id 0x%lx existing-bp 0x%lx new-bp:0x%lx",
+                    update.entry.mac.to_string().c_str(), entry->bv_id, existing_entry->second.bridge_port_id, bridge_port_id);
              break;
         }
 
         update.add = true;
+        update.type = "dynamic";
         storeFdbEntryState(update);
+        update.port.m_fdb_count++;
+        m_portsOrch->setPort(update.port.m_alias, update.port);
+        vlan.m_fdb_count++;
+        m_portsOrch->setPort(vlan.m_alias, vlan);
 
         for (auto observer: m_observers)
         {
             observer->update(SUBJECT_TYPE_FDB_CHANGE, &update);
         }
 
         break;
-
+    }
     case SAI_FDB_EVENT_AGED:
-    case SAI_FDB_EVENT_MOVE:
+    {
+        SWSS_LOG_INFO("Received AGE event for bvid=%lx mac=%s port=%lx", entry->bv_id, update.entry.mac.to_string().c_str(), bridge_port_id);
+
+        if (!m_portsOrch->getPort(entry->bv_id, vlan))
+        {
+            SWSS_LOG_NOTICE("FdbOrch AGE notification: Failed to locate vlan port from bv_id 0x%lx", entry->bv_id);
+            return;
+        }
```

Comment:
It seems MAC age may don't care the existence of vlan and port, except static type MAC

Reply:
getPort(entry->bv_id, vlan) returns the vlan structure, which is used to update the vlan reference count at end of case SAI_FDB_EVENT_AGED:
        vlan.m_fdb_count--;
        m_portsOrch->setPort(vlan.m_alias, vlan);

Reply:
It is only a fault-tolerance for MAC AGE, MOVE and FLUSH. If we can guarantee the message order, getPort(entry->bv_id, vlan) will never return in this place. If there is a mistake or the message order is not guaranteed, when the vlan or port is destroyed, but MAC is still in m_entries, this MAC still must be removed from m_entries to avoid conflict with HW.
If vlan or port is destroyed, reference count may not to be updated.
if (vlan)
{
    vlan.m_fdb_count--;
    m_portsOrch->setPort(vlan.m_alias, vlan);
}

Reply:
Got it. I have changes code accordingly.